iroh-router/src/router.rs

```diff
@@ -0,0 +1,211 @@
+    future::{MapErr, Shared},
+    FutureExt, TryFutureExt,
+};
+use iroh_net::Endpoint;
+use tokio::task::{JoinError, JoinSet};
+use tokio_util::{sync::CancellationToken, task::AbortOnDropHandle};
+use tracing::{debug, error, warn};
+
+use crate::{ProtocolHandler, ProtocolMap};
+
+#[derive(Clone, Debug)]
+pub struct Router {
+    endpoint: Endpoint,
+    protocols: Arc<ProtocolMap>,
+    // `Node` needs to be `Clone + Send`, and we need to `task.await` in its `shutdown()` impl.
+    // So we need
+    // - `Shared` so we can `task.await` from all `Node` clones
+    // - `MapErr` to map the `JoinError` to a `String`, because `JoinError` is `!Clone`
+    // - `AbortOnDropHandle` to make sure that the `task` is cancelled when all `Node`s are dropped
+    //   (`Shared` acts like an `Arc` around its inner future).
+    task: Shared<MapErr<AbortOnDropHandle<()>, JoinErrToStr>>,
+    cancel_token: CancellationToken,
+}
+
+type JoinErrToStr = Box<dyn Fn(JoinError) -> String + Send + Sync + 'static>;
+
+impl Router {
+    pub fn builder(endpoint: Endpoint) -> RouterBuilder {
+        RouterBuilder::new(endpoint)
+    }
+
+    /// Returns a protocol handler for an ALPN.
+    ///
+    /// This downcasts to the concrete type and returns `None` if the handler registered for `alpn`
+    /// does not match the passed type.
+    pub fn get_protocol<P: ProtocolHandler>(&self, alpn: &[u8]) -> Option<Arc<P>> {
+        self.protocols.get_typed(alpn)
+    }
+
+    pub fn endpoint(&self) -> &Endpoint {
+        &self.endpoint
+    }
+
+    pub async fn shutdown(self) -> Result<()> {
+        // Trigger shutdown of the main run task by activating the cancel token.
+        self.cancel_token.cancel();
+
+        // Wait for the main task to terminate.
+        self.task.await.map_err(|err| anyhow!(err))?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct RouterBuilder {
+    endpoint: Endpoint,
+    protocols: ProtocolMap,
+}
+
+impl RouterBuilder {
+    pub fn new(endpoint: Endpoint) -> Self {
+        Self {
+            endpoint,
+            protocols: ProtocolMap::default(),
+        }
+    }
+
+    pub fn accept(mut self, alpn: Vec<u8>, handler: Arc<dyn ProtocolHandler>) -> Self {
+        self.protocols.insert(alpn, handler);
+        self
+    }
+
+    /// Returns the [`Endpoint`] of the node.
+    pub fn endpoint(&self) -> &Endpoint {
+        &self.endpoint
+    }
+
+    /// Returns a protocol handler for an ALPN.
+    ///
+    /// This downcasts to the concrete type and returns `None` if the handler registered for `alpn`
+    /// does not match the passed type.
+    pub fn get_protocol<P: ProtocolHandler>(&self, alpn: &[u8]) -> Option<Arc<P>> {
+        self.protocols.get_typed(alpn)
+    }
+
+    pub async fn spawn(self) -> Result<Router> {
+        // Update the endpoint with our alpns.
+        let alpns = self
+            .protocols
+            .alpns()
+            .map(|alpn| alpn.to_vec())
+            .collect::<Vec<_>>();
```

Comment:
I recently looked at how you have control over which ALPN gets selected: the server will pick the highest priority one that is supported by the client.  That means the order of the ALPNs here is important!

E.g. if you have both `iroh-blob-v1` and `iroh-blobs-v2` protocols you want to be able to pass `vec![b'iroh-blobs-v2', b'iroh-blobs-v1']` and not the other way around.

I think currently you can't do this, which could be problematic.

Reply:
agree that we should look into how to deal with this before finalizing the API